osu.Game/OsuGame.cs

```diff
@@ -725,6 +753,12 @@ public void PresentBeatmap(IBeatmapSetInfo beatmap, Predicate<BeatmapInfo> diffi
         /// <param name="password">The password to join the room, if any is given.</param>
         public void PresentMultiplayerMatch(Room room, string password)
         {
+            if (room.HasEnded)
+            {
+                Notifications.Post(new SimpleNotification { Text = "This multiplayer room has ended." });
+                return;
+            }
```

Comment:
Maybe a bit superfluous because this is handled server-side too:

https://github.com/ppy/osu-server-spectator/blob/b0edb077f1e9fd231ec4f22c15e3d66c3c3afa69/osu.Server.Spectator/Hubs/Multiplayer/MultiplayerHub.cs#L216-L217

but maybe good to have anyhow to avoid pulling the user out of whatever it is they're doing if there's no need to...?

Reply:
I'm probably missing something crucial, but won't this check stop users from being able to click a link in chat to view history?

Reply:
I suppose you're not, actually. This'll need special handling for realtime rooms, probably, and probably context-dependent too since this method was already used for multiplayer room invites I believe. 

Slightly annoying that.

Reply:
Idea: what if clicking this notification opened the room history page in browser or something? Maybe kills both birds with one stone in a rather elegant manner.

Reply:
I think that's an okay compromise, but is there a reason we don't just show the closed game screen in-game?

Reply:
Fair. Let's go with history via notification click.

Reply:
Done in https://github.com/ppy/osu/pull/32971/commits/9b3812210c8dce1ed13794c428e9214e164e2e00